### PR TITLE
 alts: Fix TsiSocket doWrite on outstanding handshake bytes handling

### PR DESCRIPTION
**Pull request by yihuazhang**
TsiSocket doWrite() incorrectly handles outstanding handshake bytes produced after handshake is marked as complete. This PR fixes the issue. 

Details: when the handshake is complete in `doHandshakeNextDone()`, it calls `raiseEvent(Network::ConnectionEvent::Connected)` which triggers `TsiSocket::doWrite()`. The `doWrite()` call then enters `repeatProtectAndWrite()` with outstanding handshake bytes in `raw_write_buffer_`, which violates the invariant.

### CODE REVIEW

**Review by antoniovicente — COMMENTED** [review]
Thanks for the quick debug and fix.

**Comment by yihuazhang** [comment]
Regarding test, we have `DoWriteOutstandingHandshakeData` that covers the case when we finish a handshake with outstanding handshake bytes in `raw_write_buffer_`.

**Comment by antoniovicente** [comment]
> Regarding test, we have `DoWriteOutstandingHandshakeData` that covers the case when we finish a handshake with outstanding handshake bytes in `raw_write_buffer_`.

The existing test didn't catch the regression introduced by https://github.com/envoyproxy/envoy/pull/15962, so I think we have some gaps in test coverage that we should address in this PR.

File: source/extensions/transport_sockets/alts/tsi_socket.cc [hunk 1]
```diff
@@ -167,8 +169,8 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
   // Try to write raw buffer when next call is done, even this is not in do[Read|Write] stack.
   if (raw_write_buffer_.length() > 0) {
     Network::IoResult result = raw_buffer_socket_->doWrite(raw_write_buffer_, false);
-    if (handshake_complete_ && raw_write_buffer_.length() > 0) {
-      write_buffer_contains_handshake_bytes_ = true;
+    if (handshake_complete_) {
```
Comment (antoniovicente) [inline source/extensions/transport_sockets/alts/tsi_socket.cc:172]: An error could be detected on doWrite which could cause the connection to be closed.  I think that this raiseEvent could then cause a crash if the connection is already closed.

Similarly, the call to doWrite after raiseEvent(Connected) needs to check the status of the connection. If the raiseEvent(Connected) closes the connection the doWrite may run into some trouble.  IIRC close on raiseEvent(Connected)  does happen in the case of SSL sockets.

Comment (yihuazhang) [reply source/extensions/transport_sockets/alts/tsi_socket.cc]: Which doWrite do you mean in "Similarly, the call to doWrite after raiseEvent(Connected) needs to check the status of the connection"? We do not call doWrite after raiseEvent in this PR.

**Comment by yihuazhang** [comment]
> > Regarding test, we have `DoWriteOutstandingHandshakeData` that covers the case when we finish a handshake with outstanding handshake bytes in `raw_write_buffer_`.
> 
> The existing test didn't catch the regression introduced by #15962, so I think we have some gaps in test coverage that we should address in this PR.

The reason why we did not catch the regression is `raiseEvent` is mocked in `MockTransportSocketCallbacks` and I do not know if there is an easy way to unmock it. WDYT?

**Comment by antoniovicente** [comment]
> > > Regarding test, we have `DoWriteOutstandingHandshakeData` that covers the case when we finish a handshake with outstanding handshake bytes in `raw_write_buffer_`.
> > 
> > 
> > The existing test didn't catch the regression introduced by #15962, so I think we have some gaps in test coverage that we should address in this PR.
> 
> The reason why we did not catch the regression is `raiseEvent` is mocked in `MockTransportSocketCallbacks` and I do not know if there is an easy way to unmock it. WDYT?

I recommend an integration test that uses tsi socket from the test client.  A write down this socket before handshake is complete may be able to repo the failure scenario that we saw on our internal extensions when tsi socket was updated.

**Comment by yihuazhang** [comment]
> > > > Regarding test, we have `DoWriteOutstandingHandshakeData` that covers the case when we finish a handshake with outstanding handshake bytes in `raw_write_buffer_`.
> > > 
> > > 
> > > The existing test didn't catch the regression introduced by #15962, so I think we have some gaps in test coverage that we should address in this PR.
> > 
> > 
> > The reason why we did not catch the regression is `raiseEvent` is mocked in `MockTransportSocketCallbacks` and I do not know if there is an easy way to unmock it. WDYT?
> 
> I recommend an integration test that uses tsi socket from the test client. A write down this socket before handshake is complete may be able to repo the failure scenario that we saw on our internal extensions when tsi socket was updated.

Nice suggestion! Let me play around with the integration test to see if we can reproduce it.

File: source/extensions/transport_sockets/alts/tsi_socket.cc [hunk 2]
```diff
@@ -266,8 +268,8 @@ Network::IoResult TsiSocket::doRead(Buffer::Instance& buffer) {
 Network::IoResult TsiSocket::repeatProtectAndWrite(Buffer::Instance& buffer, bool end_stream) {
   uint64_t total_bytes_written = 0;
   Network::IoResult result = {Network::PostIoAction::KeepOpen, 0, false};
-
-  ASSERT(!write_buffer_contains_handshake_bytes_);
+  // There should be no handshake bytes in raw_write_buffer_.
+  ASSERT(!(raw_write_buffer_.length() > 0 && prev_bytes_to_drain_ == 0));
   while (true) {
     uint64_t bytes_to_drain_this_iteration =
```
[the remark this answers is not in the document]
Comment (yihuazhang) [reply source/extensions/transport_sockets/alts/tsi_socket.cc:274]: Sure thing. Done.

**Comment by yihuazhang** [comment]
> > > > > Regarding test, we have `DoWriteOutstandingHandshakeData` that covers the case when we finish a handshake with outstanding handshake bytes in `raw_write_buffer_`.
> > > > 
> > > > 
> > > > The existing test didn't catch the regression introduced by #15962, so I think we have some gaps in test coverage that we should address in this PR.
> > > 
> > > 
> > > The reason why we did not catch the regression is `raiseEvent` is mocked in `MockTransportSocketCallbacks` and I do not know if there is an easy way to unmock it. WDYT?
> > 
> > 
> > I recommend an integration test that uses tsi socket from the test client. A write down this socket before handshake is complete may be able to repo the failure scenario that we saw on our internal extensions when tsi socket was updated.
> 
> Nice suggestion! Let me play around with the integration test to see if we can reproduce it.

@antoniovicente, I tried to add some data to client's buffer before starting an ALTS connection, and it seems one place where we can add [write() API](https://github.com/envoyproxy/envoy/blob/main/source/common/network/connection_impl.cc#L437) is before [this API] (https://github.com/envoyproxy/envoy/blob/main/test/integration/http_integration.cc#L271-L272) returns in makeRawHttpConnection() in http_integration.cc

Then, I got stuck on how to add a partial HTTP request via the above write API because it seems an HTTP request is encoded via an [RequestEncoder]() which does not provide an API that returns the final string to be sent to the peer. In other words, it does not provide an API that we can use to manipulate the HTTP request. Ideally, if there are APIs that can send/receive a raw string over the ALTS connection, we can easily use them to test the above regression. Could you please provide some guidance on it?

**Comment by antoniovicente** [comment]
> > > > > > Regarding test, we have `DoWriteOutstandingHandshakeData` that covers the case when we finish a handshake with outstanding handshake bytes in `raw_write_buffer_`.
> > > > > 
> > > > > 
> > > > > The existing test didn't catch the regression introduced by #15962, so I think we have some gaps in test coverage that we should address in this PR.
> > > > 
> > > > 
> > > > The reason why we did not catch the regression is `raiseEvent` is mocked in `MockTransportSocketCallbacks` and I do not know if there is an easy way to unmock it. WDYT?
> > > 
> > > 
> > > I recommend an integration test that uses tsi socket from the test client. A write down this socket before handshake is complete may be able to repo the failure scenario that we saw on our internal extensions when tsi socket was updated.
> > 
> > 
> > Nice suggestion! Let me play around with the integration test to see if we can reproduce it.
> 
> @antoniovicente, I tried to add some data to client's buffer before starting an ALTS connection, and it seems one place where we can add [write() API](https://github.com/envoyproxy/envoy/blob/main/source/common/network/connection_impl.cc#L437) is before [this API](https://github.com/envoyproxy/envoy/blob/main/test/integration/http_integration.cc#L271-L272) returns in makeRawHttpConnection() in http_integration.cc
> 
> Then, I got stuck on how to add a partial HTTP request via the above write API because it seems an HTTP request is encoded via an RequestEncoder which does not provide an API that returns the final string to be sent to the peer. In other words, it does not provide an API that we can use to manipulate the HTTP request. Ideally, if there are APIs that can send/receive a raw string over the ALTS connection, we can easily use them to test the above regression. Could you please provide some guidance on it?

BaseIntegrationTest::sendRawHttpAndWaitForResponse seems like a possible starting point.  The RawConnectionDriver does a write when onEvent(Network::ConnectionEvent::Connected) is delivered by the transport.

**Comment by yihuazhang** [comment]
> BaseIntegrationTest::sendRawHttpAndWaitForResponse seems like a possible starting point. The RawConnectionDriver does a write when onEvent(Network::ConnectionEvent::Connected) is delivered by the transport.

I added a test that uses `sendRawHttpAndWaitForResponse`, but it just times out. Do you have any suggestion on what I did wrong? (the most recent commit has the added test).

**Comment by antoniovicente** [comment]
> > BaseIntegrationTest::sendRawHttpAndWaitForResponse seems like a possible starting point. The RawConnectionDriver does a write when onEvent(Network::ConnectionEvent::Connected) is delivered by the transport.
> 
> I added a test that uses `sendRawHttpAndWaitForResponse`, but it just times out. Do you have any suggestion on what I did wrong? (the most recent commit has the added test).

Try applying the attached patch using 'git am'

[fix-AltsIntegrationTestValidPeer.RouterRequestAndRes.txt](https://github.com/envoyproxy/envoy/files/6525524/fix-AltsIntegrationTestValidPeer.RouterRequestAndRes.txt)

**Comment by yihuazhang** [comment]
> > > BaseIntegrationTest::sendRawHttpAndWaitForResponse seems like a possible starting point. The RawConnectionDriver does a write when onEvent(Network::ConnectionEvent::Connected) is delivered by the transport.
> > 
> > 
> > I added a test that uses `sendRawHttpAndWaitForResponse`, but it just times out. Do you have any suggestion on what I did wrong? (the most recent commit has the added test).
> 
> Try applying the attached patch using 'git am'
> 
> [fix-AltsIntegrationTestValidPeer.RouterRequestAndRes.txt](https://github.com/envoyproxy/envoy/files/6525524/fix-AltsIntegrationTestValidPeer.RouterRequestAndRes.txt)

Thanks much for the patch, @antoniovicente! The test passes now! I also verified that the same test fails in the old implementation.

**Comment by antoniovicente** [comment]
> > > > BaseIntegrationTest::sendRawHttpAndWaitForResponse seems like a possible starting point. The RawConnectionDriver does a write when onEvent(Network::ConnectionEvent::Connected) is delivered by the transport.
> > > 
> > > 
> > > I added a test that uses `sendRawHttpAndWaitForResponse`, but it just times out. Do you have any suggestion on what I did wrong? (the most recent commit has the added test).
> > 
> > 
> > Try applying the attached patch using 'git am'
> > [fix-AltsIntegrationTestValidPeer.RouterRequestAndRes.txt](https://github.com/envoyproxy/envoy/files/6525524/fix-AltsIntegrationTestValidPeer.RouterRequestAndRes.txt)
> 
> Thanks much for the patch, @antoniovicente! The test passes now! I also verified that the same test fails in the old implementation. So in my previous commit, `makeAltsConnection()` that gets passed to `makeHttpConnection()` whose result is assigned to `codec_client_` does not actually take effect or at least, is not a correct way of establishing a connection?

The path used when sending raw bytes on the connection doesn't use the codec.  I think that your prior attempt ended up creating 2 connections to the proxy, but only 1 of the two used Alts, the other was a raw TCP connection so we ended up sending an HTTP request instead of an ALTS handshake to the ALTS listener.

**Comment by antoniovicente** [comment]
Please look at the `precheck format` CI.  I forgot to run format before sending you that patch.

/wait

**Comment by yihuazhang** [comment]
> The path used when sending raw bytes on the connection doesn't use the codec. I think that your prior attempt ended up creating 2 connections to the proxy, but only 1 of the two used Alts, the other was a raw TCP connection so we ended up sending an HTTP request instead of an ALTS handshake to the ALTS listener.

Ah it makes sense. It could have taken me a while to figure this out.  +1 for adding the check to ensure codecs are not created when using raw connections.

**Comment by antoniovicente** [comment]
I'm not sure why tests don't seem to be running.  Could you merge upstream/main ?

**Comment by antoniovicente** [comment]
Please look at the TSAN/ASAN CI results.  There seems to be a test-only bug in AltsIntegrationTestValidPeer.RouterRequestAndResponseWithBodyRawHttp which results in use-after-free.

**Comment by yihuazhang** [comment]
The use-after-free error happened when invoking `verifyActualFrameSizeToUse` on `client_tsi_socket_`, which could be nullptr since we already transferred its ownership to `createClientConnection()`.

**Comment by yihuazhang** [comment]
@antoniovicente Do you know the version of gRPC envoy currently depends on? I wonder if it includes https://github.com/grpc/grpc/pull/25621. If it does, we can verify the correctness of [`max_frame_size`](https://github.com/grpc/grpc/blob/master/test/core/tsi/alts/fake_handshaker/handshaker.proto#L100) in the same way we verify other fields (i.e., [rpc_versions](https://github.com/envoyproxy/envoy/blob/main/test/extensions/transport_sockets/alts/alts_integration_test.cc#L63)) in the client_start and server_start handshake messages. Although it will be a little different in a sense that we currently verify the negotiated frame size (i.e., `actual_frame_size_to_use_`) while the aforementioned test verifies `max_frame_size` provided in a handshake message. 

I can not think of a better way of verifying `actual_frame_size_to_use_` other than creating a static member variable in `tsi_socket` class that stores `actual_frame_size_to_use_` and verify its value in the test (need to reset it for each test case). It is pretty ugly, and I prefer not do so. 

So if the current gRPC envoy depends on does not include https://github.com/grpc/grpc/pull/25621, I prefer to add a TODO and add the aforementioned test once the PR gets included. We should also remove the use of `verifyActualFrameSizeToUse()` in the tests since it has the problem of use-after-free. 

PLMK what you think.

**Comment by antoniovicente** [comment]
You can see the current grpc dependency [here.](https://github.com/envoyproxy/envoy/blob/6ada505c76753bae978ed279851f9789ce808f08/bazel/repository_locations.bzl#L242)  We're currently building against a grpc library that is 6 months old.  I'm trying to get the dependency updated.

**Comment by yihuazhang** [comment]
SG. PLMK when the update is complete, and I will update the test.

**Comment by antoniovicente** [comment]
grpc version PR: https://github.com/envoyproxy/envoy/pull/16687

**Comment by antoniovicente** [comment]
The PR to update grpc deps just landed, please `git merge upstream/main` and look into addressing the remaining tsan/asan issues.

**Comment by antoniovicente** [comment]
/retest

since this PR doesn't change docs inputs.

**Review by antoniovicente — APPROVED** [review]
Thanks!

**Pull request merged**